source/extensions/filters/http/cache/cache_headers_utils.cc

```diff
@@ -168,6 +168,100 @@ absl::optional<uint64_t> CacheHeadersUtils::readAndRemoveLeadingDigits(absl::str
   return absl::nullopt;
 }
 
+absl::flat_hash_set<std::string> VaryHeader::parseAllowlist() {
+  // TODO(cbdm): Populate the hash_set from
+  // envoy::extensions::filters::http::cache::v3alpha::CacheConfig::allowed_vary_headers.
+  // Need to make sure that the headers we add here are valid values (i.e., not malformed). That
+  // way, we won't have to check this again in isAllowed.
+  return {"accept-encoding"};
+}
+
+bool VaryHeader::isAllowed(const absl::flat_hash_set<std::string>& allowed_headers,
+                           const Http::ResponseHeaderMap& headers) {
+  if (!hasVary(headers)) {
+    return true;
+  }
+
+  std::vector<std::string> varied_headers =
+      parseHeaderValue(headers.get(Http::Headers::get().Vary));
+
+  // If the vary value was malformed, it will not be contained in allowed_headers.
+  for (const std::string& header : varied_headers) {
+    if (!allowed_headers.contains(header)) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+bool VaryHeader::hasVary(const Http::ResponseHeaderMap& headers) {
+  const Http::HeaderEntry* vary_header = headers.get(Http::Headers::get().Vary);
+  return ((vary_header) && (!vary_header->value().empty()));
+}
+
+std::string VaryHeader::createVaryKey(const Http::HeaderEntry* vary_header,
+                                      const Http::RequestHeaderMapPtr& entry_headers) {
+  if (!vary_header) {
+    return "";
+  }
+
+  ASSERT(vary_header->key() == "vary");
+
+  std::string vary_key = "vary-key\n";
+
+  for (const std::string& header : parseHeaderValue(vary_header)) {
+    // TODO(cbdm): Can add some bucketing logic here based on header. For example, we could
+    // normalize the values for accept-language by making all of {en-CA, en-GB, en-US} into
+    // "en". This way we would not need to store multiple versions of the same payload, and any
+    // of those values would find the payload in the requested language. The config should enable
+    // and control the bucketing wanted.
```

Comment:
FYI another example would be try to try to suss out iphone/android/desktop in the UA and offer that as a course-grained cache variance.  There's some hack code for that already in source/common/http/user_agent.cc.

FYI there's a bunch of other bucketing in another project here: https://github.com/apache/incubator-pagespeed-mod/blob/master/pagespeed/kernel/http/user_agent_matcher.cc which might give you an idea of the kinds of things that might be useful in bucketing. It's a messy topic.

Reply:
Thanks for the pointers!
I expanded the comment to mention `user_agent.cc` as code that could be used as a base for some bucketing logic.